File: packages/typescript/src/core/limiter.ts
/**
 * Main rate limiter implementation.
 */

import { Decision } from './decision';
import { Policy, KeyStrategy, Algorithm, normalizePolicy } from './policy';
import {
    extractIp,
    extractUserId,
    extractApiKey,
    extractPath,
    isHealthCheck,
    isPrivateIp,
} from './extractors';
import { TokenBucket } from '../algorithms/token-bucket';
import { FixedWindow } from '../algorithms/fixed-window';
import { SlidingWindow } from '../algorithms/sliding-window';
import { LeakyBucket } from '../algorithms/leaky-bucket';
import { Store } from '../stores/memory';
import { AtomicStore, isAtomicStore } from './store';
import { TelemetryHooks } from './telemetry';

type AlgorithmInstance = TokenBucket | FixedWindow | SlidingWindow | LeakyBucket;

export interface RateLimiterOptions {
    /**
     * Storage backend. Either a simple KV store (InMemoryStore — the limiter runs
     * the algorithm in-process) or an atomic store (RedisStore — the store
     * computes the decision atomically via Lua).
     */
    store: Store | AtomicStore;
    /** Either a static policy or a resolver returning a policy per-request */
    policy: Policy | ((request: any) => Policy | Promise<Policy>);
    trustedProxies?: string[];
    exemptPrivateIps?: boolean;
    /** Optional OpenTelemetry-like tracer (object with startSpan) */
    otelTracer?: any;
    /** Optional metrics recorder: (name, tags?, value?) => void */
    metricsRecorder?: (name: string, tags?: Record<string, string>, value?: number) => void;
    /**
     * Optional high-level observability hooks (StatsCollector, OpenTelemetryMetrics,
     * LoggingTelemetry, or a CompositeTelemetry of several). Called on every
     * rate-limited check with rich metadata (policy, algorithm, endpoint, cost, plan).
     */
    telemetry?: TelemetryHooks;
}

export class RateLimiter {
    private store: Store | AtomicStore;
    private policyOrResolver: Policy | ((request: any) => Policy | Promise<Policy>);
    private trustedProxies: string[];
    private exemptPrivateIps: boolean;
    private algorithmCache: Map<string, AlgorithmInstance>;
    private otelTracer?: any;
    private metricsRecorder?: (name: string, tags?: Record<string, string>, value?: number) => void;
    private telemetry?: TelemetryHooks;

    constructor(options: RateLimiterOptions) {
        this.store = options.store;
        this.policyOrResolver = options.policy;
        this.trustedProxies = options.trustedProxies ?? [];
        this.exemptPrivateIps = options.exemptPrivateIps ?? true;
        this.algorithmCache = new Map();
        this.otelTracer = options.otelTracer;
        this.metricsRecorder = options.metricsRecorder;
        this.telemetry = options.telemetry;
    }

    /** Fan a finished decision out to the telemetry hooks with rich metadata. */
    private emitTelemetry(
        request: any,
        key: string,
        decision: Decision,
        policy: Required<Policy>,
        cost: number
    ): void {
        if (!this.telemetry) return;
        const metadata: Record<string, any> = {
            policy: policy.name,
            algorithm: policy.algorithm,
            keyStrategy: policy.keyStrategy,
            endpoint: extractPath(request) ?? undefined,
            cost,
            plan: policy.plan,
        };
        this.telemetry.onCheck?.(key, decision, metadata);
        if (decision.allowed) {
            this.telemetry.onAllowed?.(key, decision, metadata);
        } else {
            this.telemetry.onBlocked?.(key, decision, metadata);
        }
    }

    /**
     * Check if request is allowed under rate limit.
     */
    /**
     * Check if request is allowed under rate limit.
     * This method is async because policy resolution may be async (e.g. DB lookup).
     */
    async check(request: any, cost?: number): Promise<Decision> {
        // Resolve policy (static or per-request)
        const resolved =
            typeof this.policyOrResolver === 'function'
                ? await (this.policyOrResolver as (r: any) => Policy | Promise<Policy>)(request)
                : (this.policyOrResolver as Policy);

        const policy = normalizePolicy(resolved);
        const requestCost = cost ?? policy.cost;

        // Quick exemptions check (policy-aware)
        if (this.isExempt(request, policy)) {
            const resp: Decision = {
                allowed: true,
                limit: policy.limit,
                remaining: policy.limit,
                resetAt: Math.floor(Date.now() / 1000 + policy.window),
            };
            this.metricsRecorder?.('halt.request.exempt', { policy: policy.name }, 1);
            return resp;
        }

        // Extract key
        const key = this.extractKey(request, policy);
        if (!key) {
            const resp: Decision = {
                allowed: true,
                limit: policy.limit,
                remaining: policy.limit,
                resetAt: Math.floor(Date.now() / 1000 + policy.window),
            };
            this.metricsRecorder?.('halt.request.no_key', { policy: policy.name }, 1);
            return resp;
        }

        const storageKey = `halt:${policy.name}:${key}`;

        // Atomic stores (e.g. Redis) compute the whole decision themselves.
        if (isAtomicStore(this.store)) {
            const span = this.otelTracer?.startSpan?.('halt.check', {
                attributes: { policy: policy.name, key },
            });
            const decision = await this.store.evaluate({
                key: storageKey,
                algorithm: policy.algorithm,
                limit: policy.limit,
                window: policy.window,
                burst: policy.burst,
                cost: requestCost,
                ttl: policy.window * 2,
            });
            this.metricsRecorder?.(
                'halt.request.checked',
                { policy: policy.name, allowed: String(decision.allowed) },
                1
            );
            this.metricsRecorder?.(
                decision.allowed ? 'halt.request.allowed' : 'halt.request.blocked',
                { policy: policy.name },
                1
            );
            span?.end?.();
            this.emitTelemetry(request, key, decision, policy, requestCost);
            return decision;
        }

        const store = this.store as Store;

        // Get or create algorithm instance for this policy. The cache key includes
        // the parameters that affect behavior, so changing a policy's limit/window
        // at runtime (dynamic limits) takes effect immediately on the in-app path.
        const cacheKey = `${policy.name}|${policy.algorithm}|${policy.limit}|${policy.window}|${policy.burst}`;
        let algorithm = this.algorithmCache.get(cacheKey);
        if (!algorithm) {
            if (policy.algorithm === Algorithm.TOKEN_BUCKET) {
                algorithm = new TokenBucket(policy.burst, policy.limit, policy.window);
            } else if (policy.algorithm === Algorithm.FIXED_WINDOW) {
                algorithm = new FixedWindow(policy.limit, policy.window);
            } else if (policy.algorithm === Algorithm.SLIDING_WINDOW) {
                algorithm = new SlidingWindow(policy.limit, policy.window);
            } else if (policy.algorithm === Algorithm.LEAKY_BUCKET) {
                const leakRate = policy.limit / policy.window;
                algorithm = new LeakyBucket(policy.burst, leakRate, policy.window);
            } else {
                throw new Error(`Algorithm ${policy.algorithm} not implemented`);
            }
            this.algorithmCache.set(cacheKey, algorithm);
        }

        // Instrumentation: start a span if tracer available
        const span = this.otelTracer?.startSpan?.('halt.check', { attributes: { policy: policy.name, key } });

        const state = store.get(storageKey);
        let decision: Decision;

        if (algorithm instanceof TokenBucket) {
            let tokens: number;
            let lastRefill: number;

            if (!state) {
                const initialState = algorithm.initialState();
                tokens = initialState.tokens;
                lastRefill = initialState.lastRefill;
            } else {
                tokens = state.tokens;
                lastRefill = state.lastRefill;
            }

            const result = algorithm.checkAndConsume(tokens, lastRefill, requestCost);
            decision = result.decision;

            const ttl = policy.window * 2;
            store.set(storageKey, { tokens: result.newTokens, lastRefill: result.newLastRefill }, ttl);
        } else if (algorithm instanceof FixedWindow) {
            let count: number;
            let windowStart: number;

            if (!state) {
                const initialState = algorithm.initialState();
                count = initialState.count;
                windowStart = initialState.windowStart;
            } else {
                count = state.count;
                windowStart = state.windowStart;
            }

            const result = algorithm.checkAndConsume(count, windowStart, requestCost);
            decision = result.decision;

            const ttl = policy.window * 2;
            store.set(storageKey, { count: result.newCount, windowStart: result.newWindowStart }, ttl);
        } else if (algorithm instanceof SlidingWindow) {
            const buckets = state || algorithm.initialState();
            const result = algorithm.checkAndConsume(buckets, requestCost);
            decision = result.decision;

            const ttl = policy.window * 2;
            store.set(storageKey, result.newBuckets, ttl);
        } else if (algorithm instanceof LeakyBucket) {
            let level: number;
            let lastLeak: number;

            if (!state) {
                const initialState = algorithm.initialState();
                level = initialState.level;
                lastLeak = initialState.lastLeak;
            } else {
                level = state.level;
                lastLeak = state.lastLeak;
            }

            const result = algorithm.checkAndConsume(level, lastLeak, requestCost);
            decision = result.decision;

            const ttl = policy.window * 2;
            store.set(storageKey, { level: result.newLevel, lastLeak: result.newLastLeak }, ttl);
        } else {
            throw new Error(`Algorithm ${typeof algorithm} not supported`);
        }

        // Metrics and telemetry
        this.metricsRecorder?.('halt.request.checked', { policy: policy.name, allowed: String(decision.allowed) }, 1);
        if (decision.allowed) {
            this.metricsRecorder?.('halt.request.allowed', { policy: policy.name }, 1);
        } else {
            this.metricsRecorder?.('halt.request.blocked', { policy: policy.name }, 1);
        }

        span?.end?.();

        this.emitTelemetry(request, key, decision, policy, requestCost);

        return decision;
    }

    /**
     * Extract rate limit key from request based on policy strategy.
     */
    private extractKey(request: any, policy: Policy): string | null {
        // Use custom extractor if provided
        if (policy.keyExtractor) {
            return policy.keyExtractor(request);
        }

        // Use built-in strategies
        if (policy.keyStrategy === KeyStrategy.IP) {
            return extractIp(request, this.trustedProxies);
        }

        if (policy.keyStrategy === KeyStrategy.USER) {
            return extractUserId(request);
        }

        if (policy.keyStrategy === KeyStrategy.API_KEY) {
            return extractApiKey(request);
        }

        if (policy.keyStrategy === KeyStrategy.COMPOSITE) {
            // Composite: user:ip or api_key:ip
            const user = extractUserId(request);
            const apiKey = extractApiKey(request);
            const ip = extractIp(request, this.trustedProxies);

            if (user && ip) {
                return `${user}:${ip}`;
            } else if (apiKey && ip) {
                return `${apiKey}:${ip}`;
            } else if (user) {
                return user;
            } else if (apiKey) {
                return apiKey;
            } else {
                return ip;
            }
        }

        return null;
    }

    /**
     * Check if request is exempt from rate limiting.
     */
    private isExempt(request: any, policy: Policy): boolean {
        // Check health check paths
        const path = extractPath(request);
        if (path && isHealthCheck(path)) {
            return true;
        }

        // Check custom exemptions
        if (path && (policy.exemptions ?? []).includes(path)) {
            return true;
        }

        // Check private IPs
        if (this.exemptPrivateIps) {
            const ip = extractIp(request, this.trustedProxies);
            if (ip && isPrivateIp(ip)) {
                return true;
            }
        }

        // Check IP exemptions
        const ip = extractIp(request, this.trustedProxies);
        if (ip && (policy.exemptions ?? []).includes(ip)) {
            return true;
        }

        return false;
    }
}

File: packages/python/halt/core/limiter.py
"""Main rate limiter implementation."""

import inspect
import time
from typing import Any, Callable, Optional, Tuple, Union
from halt.core.policy import Policy, KeyStrategy, Algorithm
from halt.core.decision import Decision
from halt.core.extractors import (
    extract_ip,
    extract_user_id,
    extract_api_key,
    is_health_check,
    is_private_ip,
)
from halt.algorithms.token_bucket import TokenBucket
from halt.algorithms.fixed_window import FixedWindow
from halt.algorithms.sliding_window import SlidingWindow
from halt.algorithms.leaky_bucket import LeakyBucket


class RateLimiter:
    """Main rate limiter class that orchestrates policy, storage, and algorithms."""

    def __init__(
        self,
        store: Any,
        policy: "Union[Policy, Callable[[Any], Policy]]",
        trusted_proxies: Optional[list] = None,
        exempt_private_ips: bool = True,
        otel_tracer: Optional[Any] = None,
        metrics_recorder: Optional[Callable] = None,
        telemetry: Optional[Any] = None,
    ) -> None:
        """Initialize rate limiter.

        Args:
            store: Storage backend. A simple store (InMemoryStore — the limiter
                runs the algorithm in-process) or an atomic store (RedisStore /
                AsyncRedisStore — the store computes the decision via Lua).
            policy: A static Policy or a resolver callable returning a Policy
                per request (for per-user / per-plan limits).
            trusted_proxies: Trusted proxy IPs/networks for X-Forwarded-For.
            exempt_private_ips: Whether to exempt private IPs from rate limiting.
            otel_tracer: Optional OpenTelemetry-like tracer (has ``start_span``).
            metrics_recorder: Optional ``(name, tags, value)`` metrics hook.
            telemetry: Optional high-level observability hooks (StatsCollector,
                OpenTelemetryMetrics, LoggingTelemetry, or a CompositeTelemetry).
        """
        self.store = store
        self.policy_or_resolver = policy
        self.trusted_proxies = trusted_proxies or []
        self.exempt_private_ips = exempt_private_ips
        self.otel_tracer = otel_tracer
        self.metrics_recorder = metrics_recorder
        self.telemetry = telemetry
        # Algorithm instances are built lazily per resolved policy and cached
        # (only used for the in-process path; atomic stores run Lua instead).
        self._algorithm_cache: dict = {}

    def _emit_telemetry(
        self, request: Any, key: str, decision: Decision, policy: Policy, cost: int
    ) -> None:
        """Fan a finished decision out to the telemetry hooks with rich metadata."""
        if not self.telemetry:
            return
        algorithm = policy.algorithm
        metadata = {
            "policy": policy.name,
            "algorithm": algorithm.value if hasattr(algorithm, "value") else algorithm,
            "key_strategy": getattr(policy.key_strategy, "value", policy.key_strategy),
            "endpoint": self._get_path(request),
            "cost": cost,
            "plan": policy.plan,
        }
        self.telemetry.on_check(key, decision, metadata)
        if decision.allowed:
            self.telemetry.on_allowed(key, decision, metadata)
        else:
            self.telemetry.on_blocked(key, decision, metadata)

    # ----- algorithm construction (in-process path) -----------------------

    def _get_algorithm(self, policy: Policy) -> Any:
        """Build (and cache) the algorithm instance for a resolved policy.

        Cached by the parameters that affect behavior, so a resolver returning
        different limits under the same policy name stays correct.
        """
        cache_key = (
            policy.name,
            policy.algorithm,
            policy.limit,
            policy.window,
            policy.burst,
            policy.sliding_precision,
        )
        algorithm = self._algorithm_cache.get(cache_key)
        if algorithm is not None:
            return algorithm

        if policy.algorithm == Algorithm.TOKEN_BUCKET:
            algorithm = TokenBucket(
                capacity=policy.burst or policy.limit,
                rate=policy.limit,
                window=policy.window,
            )
        elif policy.algorithm == Algorithm.FIXED_WINDOW:
            algorithm = FixedWindow(limit=policy.limit, window=policy.window)
        elif policy.algorithm == Algorithm.SLIDING_WINDOW:
            algorithm = SlidingWindow(
                limit=policy.limit,
                window=policy.window,
                precision=policy.sliding_precision,
            )
        elif policy.algorithm == Algorithm.LEAKY_BUCKET:
            leak_rate = policy.limit / policy.window
            algorithm = LeakyBucket(
                capacity=policy.burst or policy.limit,
                leak_rate=leak_rate,
                window=policy.window,
            )
        else:
            raise NotImplementedError(f"Algorithm {policy.algorithm} not implemented")

        self._algorithm_cache[cache_key] = algorithm
        return algorithm

    # ----- shared request preparation -------------------------------------

    def _allow_all(self, policy: Policy) -> Decision:
        return Decision(
            allowed=True,
            limit=policy.limit,
            remaining=policy.limit,
            reset_at=int(time.time() + policy.window),
        )

    def _resolve_policy(self, request: Any) -> Any:
        """Resolve the policy (may return a coroutine for async resolvers)."""
        if callable(self.policy_or_resolver):
            return self.policy_or_resolver(request)
        return self.policy_or_resolver

    def _prepare(
        self, request: Any, policy: Policy, cost: Optional[int]
    ) -> Tuple[Optional[Decision], Optional[Policy], Optional[str], Optional[str], int]:
        """Apply exemptions and extract the storage key for an already-resolved policy.

        Returns ``(early_decision, policy, storage_key, raw_key, cost)``. When
        ``early_decision`` is set the caller should return it directly.
        """
        if cost is None:
            cost = policy.cost

        if self._is_exempt(request, policy):
            return self._allow_all(policy), policy, None, None, cost

        key = self._extract_key(request, policy)
        if key is None:
            return self._allow_all(policy), policy, None, None, cost

        storage_key = f"halt:{policy.name}:{key}"
        return None, policy, storage_key, key, cost

    def _evaluate_args(self, policy: Policy, storage_key: str, cost: int) -> dict:
        return dict(
            key=storage_key,
            algorithm=policy.algorithm,
            limit=policy.limit,
            window=policy.window,
            burst=policy.burst or policy.limit,
            cost=cost,
            ttl=policy.window * 2,
        )

    # ----- public API -----------------------------------------------------

    def check(self, request: Any, cost: Optional[int] = None) -> Decision:
        """Check if a request is allowed under the rate limit (synchronous).

        Delegates to an atomic store's ``evaluate`` when available, otherwise
        runs the algorithm in-process against the simple store.
        """
        policy = self._resolve_policy(request)
        if inspect.iscoroutine(policy):
            policy.close()
            raise RuntimeError(
                "Async policy resolver requires acheck(): use `await limiter.acheck(request)`"
            )
        early, policy, storage_key, raw_key, cost = self._prepare(request, policy, cost)
        if early is not None:
            return early

        if hasattr(self.store, "evaluate"):
            decision = self.store.evaluate(**self._evaluate_args(policy, storage_key, cost))
        else:
            decision = self._compute_in_app(policy, storage_key, cost)

        self._emit_telemetry(request, raw_key, decision, policy, cost)
        return decision

    async def acheck(self, request: Any, cost: Optional[int] = None) -> Decision:
        """Async variant of :meth:`check` for async stores / frameworks.

        Uses an atomic store's ``aevaluate`` when available, falls back to a
        synchronous ``evaluate``, and finally to the in-process path. Supports
        async policy resolvers (e.g. a cached loader reading Redis/DB).
        """
        policy = self._resolve_policy(request)
        if inspect.iscoroutine(policy):
            policy = await policy
        early, policy, storage_key, raw_key, cost = self._prepare(request, policy, cost)
        if early is not None:
            return early

        if hasattr(self.store, "aevaluate"):
            decision = await self.store.aevaluate(**self._evaluate_args(policy, storage_key, cost))
        elif hasattr(self.store, "evaluate"):
            decision = self.store.evaluate(**self._evaluate_args(policy, storage_key, cost))
        else:
            decision = self._compute_in_app(policy, storage_key, cost)

        self._emit_telemetry(request, raw_key, decision, policy, cost)
        return decision

    # ----- in-process algorithm path --------------------------------------

    def _compute_in_app(self, policy: Policy, storage_key: str, cost: int) -> Decision:
        """Run the algorithm in-process using the simple (KV) store."""
        algorithm = self._get_algorithm(policy)
        state = self.store.get(storage_key)
        ttl = policy.window * 2

        if isinstance(algorithm, TokenBucket):
            if state is None:
                tokens, last_refill = algorithm.initial_state()
            else:
                tokens, last_refill = state
            decision, new_tokens, new_last_refill = algorithm.check_and_consume(
                current_tokens=tokens, last_refill=last_refill, cost=cost
            )
            self.store.set(storage_key, (new_tokens, new_last_refill), ttl=ttl)

        elif isinstance(algorithm, FixedWindow):
            if state is None:
                count, window_start = algorithm.initial_state()
            else:
                count, window_start = state
            decision, new_count, new_window_start = algorithm.check_and_consume(
                current_count=count, window_start=window_start, cost=cost
            )
            self.store.set(storage_key, (new_count, new_window_start), ttl=ttl)

        elif isinstance(algorithm, SlidingWindow):
            buckets = state if state is not None else algorithm.initial_state()
            decision, new_buckets = algorithm.check_and_consume(buckets=buckets, cost=cost)
            self.store.set(storage_key, new_buckets, ttl=ttl)

        elif isinstance(algorithm, LeakyBucket):
            if state is None:
                level, last_leak = algorithm.initial_state()
            else:
                level, last_leak = state
            decision, new_level, new_last_leak = algorithm.check_and_consume(
                current_level=level, last_leak=last_leak, cost=cost
            )
            self.store.set(storage_key, (new_level, new_last_leak), ttl=ttl)

        else:
            raise NotImplementedError(f"Algorithm {type(algorithm)} not supported")

        return decision

    # ----- key extraction & exemptions ------------------------------------

    def _extract_key(self, request: Any, policy: Policy) -> Optional[str]:
        """Extract the rate limit key from a request based on policy strategy."""
        if policy.key_extractor:
            return policy.key_extractor(request)

        if policy.key_strategy == KeyStrategy.IP:
            return extract_ip(request, self.trusted_proxies)

        elif policy.key_strategy == KeyStrategy.USER:
            return extract_user_id(request)

        elif policy.key_strategy == KeyStrategy.API_KEY:
            return extract_api_key(request)

        elif policy.key_strategy == KeyStrategy.COMPOSITE:
            user = extract_user_id(request)
            api_key = extract_api_key(request)
            ip = extract_ip(request, self.trusted_proxies)

            if user and ip:
                return f"{user}:{ip}"
            elif api_key and ip:
                return f"{api_key}:{ip}"
            elif user:
                return user
            elif api_key:
                return api_key
            else:
                return ip

        return None

    def _is_exempt(self, request: Any, policy: Policy) -> bool:
        """Check if a request is exempt from rate limiting."""
        path = self._get_path(request)
        if path and is_health_check(path):
            return True

        if path and path in policy.exemptions:
            return True

        if self.exempt_private_ips:
            ip = extract_ip(request, self.trusted_proxies)
            if ip and is_private_ip(ip):
                return True

        ip = extract_ip(request, self.trusted_proxies)
        if ip and ip in policy.exemptions:
            return True

        return False

    def _get_path(self, request: Any) -> Optional[str]:
        """Extract the request path (framework-agnostic)."""
        if hasattr(request, "url") and hasattr(request.url, "path"):
            # FastAPI/Starlette
            return request.url.path
        elif hasattr(request, "path"):
            # Flask/Django
            return request.path

        return None
